### DISCUSSION TITLE
Showing BREAD data rows based on current user role in different bread views

### DISCUSSION THREAD

**Discussion by xman4** [thedevdojo/voyager#5397] [Ideas]
I think it'd be great to add this feature to the project
to be able to select which user role can view a specific (data row/data type field) in the browse page or the add page ,.....
![Screen Shot 2021-07-06 at 5 03 56 PM](https://user-images.githubusercontent.com/4815305/124623329-31ef2800-de7c-11eb-9619-a4f7cbe56998.png)

**Comment by haexperts2020** [comment]
I like your GUI approach. I did something like this:

Add JSON in the details like this for the role:

      {
          "project_manager": {
              "edit": false
          },
      }

So in this example if the user is a 'project_manager' then this field is not display allowing that role to not edit it. They however can browse and read it. 

Then the tricky party, modify the source code:
vendor/tcg/voyager/resources/views/bread/edit-add.blade.php starting at line 60:


        @foreach($dataTypeRows as $row)
              
              @php
              
              $details = collect($row->details);
              $show = 1; // always show the field

              if(isset($details[\Auth()->user()->role->name]->edit)){ 
              if($details[\Auth()->user()->role->name]->edit == false){  
               $show = 0; // if current user matches the role and the JSON details say edit = false then lets hide it
              }}
               
               if($show == 1){
              
                  $display_options = $row->details->display ?? NULL;
                  if ($dataTypeContent->{$row->field.'_'.($edit ? 'edit' : 'add')}) {
                      $dataTypeContent->{$row->field} = $dataTypeContent->{$row->field.'_'.($edit ? 'edit' : 'add')};
                  }
              @endphp
              @if (isset($row->details->legend) && isset($row->details->legend->text))
                  <legend class="text-{{ $row->details->legend->align ?? 'center' }}" style="background-color: {{ $row->details->legend->bgcolor ?? '#f0f0f0' }};padding: 5px;">{{ $row->details->legend->text }}</legend>
              @endif

              <div class="form-group @if($row->type == 'hidden') hidden @endif col-md-{{ $display_options->width ?? 12 }} {{ $errors->has($row->field) ? 'has-error' : '' }}" @if(isset($display_options->id)){{ "id=$display_options->id" }}@endif>
                  {{ $row->slugify }}
                  <label class="control-label" for="name">{{ $row->getTranslatedAttribute('display_name') }}</label>
                  @include('voyager::multilingual.input-hidden-bread-edit-add')
                  @if (isset($row->details->view))
                      @include($row->details->view, ['row' => $row, 'dataType' => $dataType, 'dataTypeContent' => $dataTypeContent, 'content' => $dataTypeContent->{$row->field}, 'action' => ($edit ? 'edit' : 'add'), 'view' => ($edit ? 'edit' : 'add'), 'options' => $row->details])
                  @elseif ($row->type == 'relationship')
                      @include('voyager::formfields.relationship', ['options' => $row->details])
                  @else
                      {!! app('voyager')->formField($row, $dataType, $dataTypeContent) !!}
                  @endif

                  @foreach (app('voyager')->afterFormFields($row, $dataType, $dataTypeContent) as $after)
                      {!! $after->handle($row, $dataType, $dataTypeContent) !!}
                  @endforeach
                  @if ($errors->has($row->field))
                      @foreach ($errors->get($row->field) as $error)
                          <span class="help-block">{{ $error }}</span>
                      @endforeach
                  @endif
              </div>
              @php } @endphp
          @endforeach

Do something similar to the browse and read pages. Small modification with huge flexibility when it comes to roles. Something like this should be a feature in the project. I have wanted it for quite some time. Hope this helps.

**Reply by MrCrayon** [reply]
I would not advice to change the template, unless you are maintaing a fork is going to be a pain for updates.

You can do the same thing without duplicating code extending the controller and filtering dataTypeRows.
You have also to consider that hiding fields might not be enough if you don't make sure in the controller that saves data that you are not accepting those fields.

You could also create a custom model for DataType and hook something there to filter dataTypeRows, I did not try this last method and it's been a while I didn't touch voyager code so I might be wrong.
